src/uri/include/sourcemeta/jsontoolkit/uri.h

```diff
@@ -291,6 +294,9 @@ class SOURCEMETA_JSONTOOLKIT_URI_EXPORT URI {
   static auto from_fragment(std::string_view fragment) -> URI;
 
 private:
+  bool parsed = false;
```

Comment:
Do we really need to have this? As far as I can see (to which I agree) all parsing should happen on the constructors, so we can guarantee the URI is parsed when the constructor is done?

Reply:
It was my assumption before I remember that we have this issue: #811 

If we potentially re-set stuff, we need to parse it again right?

Also I noticed that on the end of this method we reparse stuff again: 
- https://github.com/sourcemeta/jsontoolkit/blob/main/src/uri/uri.cc#L295-L299
- https://github.com/sourcemeta/jsontoolkit/blob/main/src/uri/uri.cc#L270-L272

Reply:
The idea is that we only parse one. Its a bit tricky given the intermediate state with `uriparser`, but what I'm thinking about is:

- We only perform parsing at the constructors and we populate a set of members with the result. For example, we have a member that is the host, a member that is the fragment, a member that is the path, etc.
- At this point ideally we don't store the PIMPL private `uriparser` struct anymore as a member
- The setters then just interact with the pre-parsed members, never parsing again

Reply:
The ultimate goal here is to never keep a `uriparser` struct around anymore. We only do it on demand when we need to (and ideally we stop needing it at some point). For example, if we parse into a set of nice members on the class, then we might need to re-construct a `uriparser` instance on the fly for relative resolution or recomposing, and then throw it away, etc.

Reply:
> We only perform parsing at the constructors and we populate a set of members with the result. For example, we have a member that is the host, a member that is the fragment, a member that is the path, etc.

So we could parse once, then every method should not reparse but use existing members.

Which means that this should not happen?

- https://github.com/sourcemeta/jsontoolkit/blob/main/src/uri/uri.cc#L295-L299
- https://github.com/sourcemeta/jsontoolkit/blob/main/src/uri/uri.cc#L270-L272

Reply:
I'm trying to understand, how much could I apply this final vision to the refactor of URI::path which is consumed by URI::canonicalize (that reparse on the end).

Reply:
@tony-go I think its fine for the code you highlighted to happen, just that instead of affecting a `uriparser` class member directly, it would:

1. Construct back, on the fly, a `uriparser` struct based on the new class members
2. Do whatever it needs with `uriparser` to this temporary instance
3. Decode back into the new members
4. Discard the `uriparser` struct

Reply:
The core goal here is that we stop storing a `uriparser` struct as a class member, and instead put our own C++-friendly class members. Something like this:

```c++
private:
  std::optional<std::string> scheme;
  std::optional<std::string> userinfo;
  std::optional<std::string> host;
  // etc
```

That way, instead of making `uriparser` an intrinsic part of the class, we just use it on demand when needed and then discard it. That gets us closer to eventually getting rid of `uriparser`

Reply:
Yeah we are aligned. My whole point is either we add these properties and I can do uri_parse once only in the constructor.

Reply:
Yeah, let's do it!